cpp/src/community/ecg.cu

```diff
@@ -114,37 +135,51 @@ void ecg(raft::handle_t const &handle,
          vertex_t ensemble_size,
          vertex_t *clustering)
 {
+  using graph_type = GraphCSRView<vertex_t, edge_t, weight_t>;
+
   CUGRAPH_EXPECTS(graph.edge_data != nullptr,
                   "Invalid input argument: louvain expects a weighted graph");
```

Comment:
This should be - ECG only works on a weighted graph.  Is there an option to use default weights?

Reply:
From python all of the clustering algorithms will specify a default weight if no weight is included on the graph.

From C++ the user is expected to provide a weight.

Will fix the old error message in next push.

Reply:
Why `public` inheritance? If no method is overriden (virtual) then public inheritance is too much of an exposure (hence creating unnecessary coupling). Why not use `private` inheritance?

Moreover, if `Louvain` class is meant to be publicly derived (and it seems it is since it exposes at least one virtual method) then it should have a `virtual` destructor.

Reply:
Again, confusion about class naming convention, shouldn't be `<small_caps>_t`? I understand that some are older (legacy) classes, but perhaps new classed might follow the rule?

Reply:
I have added a virtual destructor in the next push.  I have clarified a few methods with `virtual` and `override` keywords to make the inheritance more clear (and possibly correct, it seems like one of the implementations might have been broken without it, although I got neither compiler errors nor erroneous results - maybe I got lucky).

EcgLouvain and Leiden are both publicly derived from Louvain.  They override different virtual methods.
`virtual weight_t operator()(size_t max_level, weight_t resolution)` is the primary entry point for all 3 classes.  EcgLouvain uses the parent version, Leiden provide its own override.  EcgLouvain overrides the `initialize_dendrogram_level` function but otherwise behaves the same as Louvain.